### DISCUSSION TITLE
Initialize variable value of newly activated element

### DISCUSSION THREAD

**Discussion by keassyguang** [idaholab/moose#29015] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello everyone,
I am currently working on a additive manufacturing process simulation project using the MOOSE. I got some problems when I tried to initialize/reinitialize variable value of newly activated elements.

According to the MOOSE documentation on ElementSubdomainModifier: https://mooseframework.inl.gov/source/meshmodifiers/ElementSubdomainModifier.html
> By default, all elements that change subdomain ID are reinitialized to the new subdomain's initial condition.

The left part of the below picture shows a temperature initialization result of a newly added layer (Layer 2) within the ABAQUS software. For the newly added layer, Layer 2, the bottom nodes of this layer are shared with Layer 1 (top nodes of layer 1). ABAQUS initializes the temperature of Layer 2 by setting the specified value, except for these shared nodes, as the temperature of these nodes is kept the same as the value at the end of Layer 1. 

As for MOOSE, based on my observations, since all elements move from the inactivated domain to the activated domain, all the nodes are initialized. For those shared nodes, it seems that MOOSE overrides their previous/existing values directly with the specified value. (If I am wrong, please correct me).

I know there might be physical meaningful for both these two initializtion methods. But what should I do if I want to achieve the same initialization as in ABAQUS? 

![image(8)](https://github.com/user-attachments/assets/a77545b8-a813-44f9-95d6-db14ec347657)

**Comment by GiudGiud** [comment]
@hugary1995 

We have not implemented this behavior afaik.
We can use the IC object to initialize, and can skip reinitialization based on the new and old subdomains these elements get assigned to. 

In the current setup, I dont think this is hard to implement. I would do it with a NeighborElementsFunctorIC, which would be an IC that looks at the neighbor element values, possibly with a multi-layer stencil, and computes the average of neighbor values within the same subdomain (the new subdomain of the element being switched)

**Reply by keassyguang** [reply]
Do you mean that I create a new IC object since I didn't find an existing IC object named "NeighborElementsFunctorIC"?

**Reply by GiudGiud** [reply]
yes

**Comment by hugary1995** [comment]
You are right about variables being initialized to their initial conditions. You can selectively initialize certain subdomains using the `reinitialize_subdomains` parameter. Quoting the parameter description

> By default, any element which changes subdomain is reinitialized. If a list of subdomains (IDs or names) is provided, then only elements whose new subdomain is in the list will be reinitialized. If an empty list is set, then no elements will be reinitialized.

As to your second question:

> As for MOOSE, based on my observations, since all elements move from the inactivated domain to the activated domain, all the nodes are initialized. For those shared nodes, it seems that MOOSE overrides their previous/existing values directly with the specified value. (If I am wrong, please correct me).

I don't believe that's the case, at least your observation is not the intended behavior. However, I see a potential issue when inspecting the source code. In `ElementSubdomainModifierBase::applyIC` method, we have the following call
```cpp
  _fe_problem.projectInitialConditionOnCustomRange(reinitializedElemRange(displaced),
                                                   reinitializedBndNodeRange(displaced));
```
It calls the `projectInitialConditionOnCustomRange` provided by the framework on the newly reinitialized elements and boundary nodes. However, in your case where you are interested in reinitializing nodal variables, this might be forcibly reinitializing the "shared" nodes (in your original description).

I agree this is an issue, and we should at least allow users to control the behavior on shared nodes.

**Reply by hugary1995** [reply]
I created #29054 to keep track of this issue. I can work on this in the coming months.

**Reply by keassyguang** [reply]
Thanks for the detailed explanation @hugary1995 and also thanks for the help so far @GiudGiud 

![image](https://github.com/user-attachments/assets/bc487c3f-3577-4f4c-9251-c3f91dab588e)


Things could get even weirder when I further checked the value of the variable at the quadrature point. As shown in the figure above, I think the current initialisation strategy might also affect the already activated elements connected to the shared nodes. Here the qp_id = 9 belongs to layer 1. The initial value of CM is around 4418, it reduces to 2323.55 after 4 seconds. However, as soon as the second layer is activated, its value changes to 4062, which should not happen as it should gradually reduce to a certain value. I also checked the `InitialConditionTempl.h`, it says:

> /** This is a template class that implements the workhorse `compute` and `computeNodal` methods. The former method is used for setting block initial conditions. It first projects the initial condition field to nodes, then to edges, then to faces, then to interior dofs. The latter `computeNodal` method sets dof values for boundary restricted initial conditions */

I'm not sure if this is also caused by shared nodes.

> I created https://github.com/idaholab/moose/issues/29054 to keep track of this issue. I can work on this in the coming months.

Hope to see this feature soon!

**Reply by GiudGiud** [reply]
> I'm not sure if this is also caused by shared nodes.

In a way. The value of a variable at a Qp is not what we actually solve for. We solve for values of variables at degrees of freedom, which can be at nodes, on sides, etc. It depends on the variable type. 
Then the variable value at Qps is computed by evaluating the shape function there. If the shape function is affected by a changed nodal value, then that could be what is happening. 

Which variable type is this?
Could you please sample the variable on the boundary instead of Qps? Before and after the subdomain change.

**Reply by keassyguang** [reply]
> Which variable type is this?

I think the variable type should be

```
order = FIRST
family = LAGRANGE
```
> Could you please sample the variable on the boundary instead of Qps? Before and after the subdomain change.

I can sample the variable on the top boundary tomorrow.

**Reply by GiudGiud** [reply]
ok so these will all be nodal DoFs.

**Reply by keassyguang** [reply]
The upper right curve shows the evolution of this variable over time. There is a sudden increase when the time equal to 4s.

**Reply by GiudGiud** [reply]
where is the "point used to plot"? Is that a node? Or a Qp? or a point on a side?

**Reply by keassyguang** [reply]
> where is the "point used to plot"? Is that a node? Or a Qp? or a point on a side?

It is a point on the top surface of layer 1, and I think it is also a node. A point on the top side of layer 1.

**Reply by GiudGiud** [reply]
@hugary1995 Is there an option to get the ICs to apply (when triggered by a subdomain change) on the new elements but not on the boundary nodes?
Or maybe this is already the default behavior?

**Reply by hugary1995** [reply]
That is the intended behavior, but it is not happening. That's exactly what I want to fix.

**Reply by GiudGiud** [reply]
Thanks. I should have re-read the entire thread!

**Reply by keassyguang** [reply]
I cannot sample the values with the `SideValueSampler` because the top_side moves from the top of layer 1 to the top of layer 2 when layer 2 is activated. So I use the `LineValueSampler` instead. On the left is time=4s and on the right is time=4.01s. The results are almost the same as before.

![Screenshot from 2024-11-14 10-04-24](https://github.com/user-attachments/assets/3a86e105-c904-4cda-8834-c9af4090d104)
